remotes/docker/scope.go

```diff
@@ -18,26 +18,93 @@ package docker
 
 import (
 	"context"
+	"fmt"
 	"net/url"
 	"sort"
 	"strings"
 
 	"github.com/containerd/containerd/reference"
 )
 
+type tokenScopes map[string]tokenScope
+
+func (scopes tokenScopes) add(ts tokenScope) {
+	match, ok := scopes[ts.resource]
+	if !ok {
+		scopes[ts.resource] = ts
+		return
+	}
+	for k := range ts.actions {
+		match.actions[k] = nil
+	}
+	scopes[ts.resource] = match
+}
+
+func (scopes tokenScopes) merge(other tokenScopes) {
+	for _, v := range other {
+		scopes.add(v)
+	}
+}
+
+func (scopes tokenScopes) contains(other tokenScopes) bool {
+	if other == nil {
```

Comment:
I think you should treat `empty` and `nil` map the same way, using `len(other)` instead.